ipv4/set_test.go

```diff
@@ -860,3 +860,148 @@ func TestOldEqualAllIPv4(t *testing.T) {
 	assert.True(t, c.Equal(a))
 	assert.True(t, c.Equal(b))
 }
+
+func TestFindPrefixWithLength(t *testing.T) {
+	tests := []struct {
+		description string
+		space       []SetI
+		reserved    []SetI
+		length      uint32
+		expected    Prefix
+		err         bool
+		change      int
+	}{
+		{
+			description: "empty",
+			space: []SetI{
+				_p("10.0.0.0/8"),
+			},
+			length: 24,
+			change: 1,
+		}, {
+			description: "find adjacent",
+			space: []SetI{
+				_p("10.0.0.0/8"),
+			},
+			reserved: []SetI{
+				_p("10.224.123.0/24"),
+			},
+			length:   24,
+			expected: _p("10.224.122.0/24"),
+		}, {
+			description: "many fewer prefixes",
+			space: []SetI{
+				_p("10.0.0.0/16"),
+			},
+			reserved: []SetI{
+				_p("10.0.1.0/24"),
+				_p("10.0.2.0/23"),
+				_p("10.0.4.0/22"),
+				_p("10.0.8.0/21"),
+				_p("10.0.16.0/20"),
+				_p("10.0.32.0/19"),
+				_p("10.0.64.0/18"),
+				_p("10.0.128.0/17"),
+			},
+			length: 24,
+			change: -7,
+		}, {
+			description: "toobig",
+			space: []SetI{
+				_p("10.0.0.0/8"),
+			},
+			reserved: []SetI{
+				_p("10.128.0.0/9"),
+				_p("10.64.0.0/10"),
+				_p("10.32.0.0/11"),
+				_p("10.16.0.0/12"),
+			},
+			length: 11,
+			err:    true,
+		}, {
+			description: "full",
+			space: []SetI{
+				_p("10.0.0.0/8"),
+			},
+			length: 7,
+			err:    true,
+		}, {
+			description: "random disjoint example",
+			space: []SetI{
+				_p("10.0.0.0/22"),
+				_p("192.168.0.0/21"),
+				_p("172.16.0.0/20"),
+			},
+			reserved: []SetI{
+				_p("192.168.0.0/21"),
+				_p("172.16.0.0/21"),
+				_p("172.16.8.0/22"),
+				_p("10.0.0.0/22"),
+				_p("172.16.12.0/24"),
+				_p("172.16.14.0/24"),
+				_p("172.16.15.0/24"),
+			},
+			length:   24,
+			expected: _p("172.16.13.0/24"),
+			change:   1,
+		},
```

Comment:
There's got to be a bunch more test cases to exercise this well.

Reply:
I think the best way to address this is to use a random number generator to pick sizes and then ensure that I can always fill up the space without ever getting a fragmentation error. Stay tuned, my next update will do this.